### PR TITLE
Implement Assembly.GetForwardedTypes() on CoreCLR

### PR DESCRIPTION
**Pull request by AtsushiKan**
Fixes https://github.com/dotnet/coreclr/issues/12391

Original approval and specs at
https://github.com/dotnet/corefx/issues/19789

### CODE REVIEW

**Comment by AtsushiKan** [comment]
@dotnet-bot test Tizen armel Cross Debug Build

File: src/mscorlib/src/System/Reflection/RuntimeAssembly.cs [hunk 1]
```diff
@@ -800,5 +801,83 @@ public override Module[] GetLoadedModules(bool getResourceModules)
 
         [MethodImplAttribute(MethodImplOptions.InternalCall)]
         internal static extern int GetToken(RuntimeAssembly assembly);
+
+        public sealed override Type[] GetForwardedTypes()
+        {
+            List<Type> types = new List<Type>();
+            List<Exception> exceptions = new List<Exception>();
```
Comment (justinvp) [inline src/mscorlib/src/System/Reflection/RuntimeAssembly.cs]: Allocate lazily?

Comment (AtsushiKan) [inline src/mscorlib/src/System/Reflection/RuntimeAssembly.cs]: It didn't seem worth the complexity given that we have two append points (one in a separate method.)

File: src/mscorlib/src/System/Reflection/RuntimeAssembly.cs [hunk 2]
```diff
@@ -800,5 +801,83 @@ public override Module[] GetLoadedModules(bool getResourceModules)
...
+                Type type = null;
+                Exception exception = null;
+                ObjectHandleOnStack pType = JitHelpers.GetObjectHandleOnStack(ref type);
+                try
+                {
+                    GetForwardedType(this, mdtExternalType, pType);
+                    if (type == null)
+                        continue;  // mdtExternalType was not a forwarder entry.
+                }
+                catch (Exception e) when (IsLoaderException(e))
```
Comment (jkotas) [inline src/mscorlib/src/System/Reflection/RuntimeAssembly.cs]: `Assembly.GetTypes` seems to be catching all exceptions. Is it intentional that we are filtering some here?

Comment (AtsushiKan) [inline src/mscorlib/src/System/Reflection/RuntimeAssembly.cs]: Yes, it's meant to mimic the CoreRT implementation where the type loading primitives reports all loader-related exceptions in a non-throwing manner.

Comment (jkotas) [inline src/mscorlib/src/System/Reflection/RuntimeAssembly.cs]: What is the behavior that callers of this API are likely to expect - consistency with existing GetTypes behavior; or consistency with internal CoreRT implemementation detail? I think it is the first one.

Comment (jkotas) [inline src/mscorlib/src/System/Reflection/RuntimeAssembly.cs]: The set of loader-related exceptions is not defined anywhere. Type loading can fail with other exceptions as well.

Comment (AtsushiKan) [inline src/mscorlib/src/System/Reflection/RuntimeAssembly.cs]: What's our fail-fast story for ExecutionEngine/AV-type exceptions that could indicate serious corrupted state?

If those are already uncatchable by managed code, I'd be a lot more comfortable with an indiscriminate catch.

Comment (jkotas) [inline src/mscorlib/src/System/Reflection/RuntimeAssembly.cs]: We do not ever throw ExecutionEngine or dirty AV exceptions - these situations are fail fasts.

Comment (AtsushiKan) [inline src/mscorlib/src/System/Reflection/RuntimeAssembly.cs]: Ok, let's remove the filter then. Please see https://github.com/dotnet/coreclr/pull/13054

**Pull request merged**